Add CCAligned Multilingual Dataset

gchhablani opened the pull request:
Hello,

I'm trying to add [CCAligned Multilingual Dataset](http://www.statmt.org/cc-aligned/). This has the potential to close #1756.

This dataset has two types - Document-Pairs, and Sentence-Pairs.

The datasets are huge, so I won't be able to test all of them. At the same time, a user might only want to download one particular language and not all. To provide this feature, `load_dataset`'s `**config_kwargs` should allow some random keyword args, in this case -`language_code`. This will be needed before the dataset is downloaded and extracted.

I'm expecting the usage to be something like - 
`load_dataset('ccaligned_multilingual','documents',language_code='en_XX-af_ZA')`. Ofcourse, at a later stage we can provide just two character language codes. This also has an issue where one language has multiple files (`my_MM` and `my_MM_zaw` on the link), but before that the required functionality must be added to `load_dataset`.

It would be great if someone could either tell me an alternative way to do this, or point me to where changes need to be made, if any, apart from the `BuilderConfig` definition. 

Additionally, I believe the tests will also have to be modified if this change is made, since it would not be possible to test for any random keyword arguments. 

A decent way to go about this would be to provide all the options in a list/dictionary for `language_code` and use that to test the arguments. In essence, this is similar to the pre-trained checkpoint dictionary as `transformers`. That means writing dataset specific tests, or adding something new to dataset generation script to make it easier for everyone to add keyword arguments without having to worry about the tests.

Thanks,
Gunjan

Requesting @lhoestq / @yjernite to review.

lhoestq commented:
Hi !

We already have some datasets that can have many many configurations possible.
To be able to support that, we allow to subclass BuilderConfig to add as many additional parameters as you may need.
This way users can load any language they want. For example the [bible_para](https://github.com/huggingface/datasets/blob/master/datasets/bible_para/bible_para.py) dataset is a dataset for translation and therefore users should be able to provide any language pair. You can check how the subclass of BuilderConfig is defined [here](https://github.com/huggingface/datasets/blob/master/datasets/bible_para/bible_para.py#L49).

For testing, only the configurations defined in the `BUILDER_CONFIGS` class attribute are used.
All the other configs combinations are not tested, but they can be used by users. If a config doesn't already exist in `BUILDER_CONFIGS`, then it is created on the fly.
For example in [bible_para](https://github.com/huggingface/datasets/blob/master/datasets/bible_para/bible_para.py#L61), only 6 configs are defined in `BUILDER_CONFIGS`.

So what I would do in your case is have something like
```python

class CCAlignedConfig(datasets.BuilderConfig):
    def __init__(self, *args, documents_or_sentences=None, language_code=None, **kwargs):
        super().__init__(
            *args,
            name=f"{documents_or_sentences}-{language_code}",
            **kwargs,
        )
        self.documents_or_sentences = documents_or_sentences
        self.language_code = language_code
```
And of course, feel free to change/rename things if you want to. In particular I think we can improve the name of the parameter `documents_or_sentences`

gchhablani commented:
Hi @lhoestq,

Thanks a lot! I don't know why I didn't think about that. :P 
I'll make these changes and update.

gchhablani commented:
Hi @lhoestq,

I have tested and added dummy files. Request you to review.

Also, does this mean BUILDER_CONFIGS is only needed while testing?

gchhablani commented:
Hi @lhoestq,

Any changes required on this one?

Thanks,
Gunjan

lhoestq commented on datasets/ccaligned_multilingual/README.md:
```diff
@@ -0,0 +1,169 @@
+annotations_creators: []
+language_creators:
+- found
+languages: []
```
Would it be possible to have the list of languages here ?

gchhablani replied:
All the languages? Or the language code given on the dataset?

lhoestq replied:
All the languages from the list on the webpage: http://www.statmt.org/cc-aligned/ :)

In terms of format we're using the BCP-47 codes. This means we use ISO-639-1 if available (`en`, `fr`, etc.). If some languages can't be formatted using an ISO-639-1 code, then you can use ISO-639-2, or ISO-639-3.
I'm pretty sure all languages for this dataset are valid ISO-639-1 code if you take what's before the `_`:
```
en_XX -> en
```

lhoestq commented on datasets/ccaligned_multilingual/README.md:
```diff
@@ -0,0 +1,169 @@
+---
+annotations_creators: []
```
```suggestion
annotations_creators:
- no-annotation
```

lhoestq commented on datasets/ccaligned_multilingual/README.md:
```diff
@@ -0,0 +1,169 @@
+- translation
+size_categories:
+- unknown
+source_datasets: []
```
```suggestion
source_datasets:
- original
```

lhoestq commented on datasets/ccaligned_multilingual/README.md:
```diff
@@ -0,0 +1,169 @@
+multilinguality:
+- translation
+size_categories:
+- unknown
```
```suggestion
- n<1K
- 1K<n<10K
- 10K<n<100K
- 100K<n<1M
- n>1M
```
let's say that there are all sizes depending on the language

lhoestq commented on datasets/ccaligned_multilingual/README.md:
```diff
@@ -0,0 +1,169 @@
+### Dataset Summary
+
+CCAligned consists of parallel or comparable web-document pairs in 137 languages aligned with English. These web-document pairs were constructed by performing language identification on raw web-documents, and ensuring corresponding language codes were corresponding in the URLs of web documents. This pattern matching approach yielded more than 100 million aligned documents paired with English. Rec ...
+
```
````suggestion

To load a language which isn't part of the config, all you need to do is specify the language code. You can find the valid languages in http://www.statmt.org/cc-aligned/ E.g.
```
dataset = load_dataset("ccaligned_multilingual", language_code="fr_XX", type="documents")
```
or
```
dataset = load_dataset("ccaligned_multilingual", language_code="fr_XX", type="sentences")
```

````
Here let's mention how to load the dataset for a specific language

lhoestq commented on datasets/ccaligned_multilingual/ccaligned_multilingual.py:
```diff
@@ -0,0 +1,213 @@
+            features = datasets.Features(
+                {
+                    "Domain": datasets.Value("string"),
+                    "Source_URL": datasets.Value("string"),
+                    "Source_Content": datasets.Value("string"),
```
Can you use the Translation feature type here please ?
```
"translation": datasets.Translation(languages=("en_XX", self.config.language_code))
```
You'll need to regenerate the dataset_infos.json file after this change

lhoestq commented on datasets/ccaligned_multilingual/ccaligned_multilingual.py:
```diff
@@ -0,0 +1,213 @@
+        else:
+            features = datasets.Features(
+                {
+                    "Source_Sentence": datasets.Value("string"),
+                    "Target_Sentence": datasets.Value("string"),
```
same here

lhoestq commented on datasets/ccaligned_multilingual/ccaligned_multilingual.py:
```diff
@@ -0,0 +1,213 @@
+        self.language_code = language_code
+
+
+class CCAlignedMultilingual(datasets.GeneratorBasedBuilder):
```
```suggestion
class CcalignedMultilingual(datasets.GeneratorBasedBuilder):
```
This must be the camel case version of the dataset script `ccaligned_multilingual`

lhoestq commented on datasets/ccaligned_multilingual/README.md:
```diff
@@ -0,0 +1,169 @@
+- [Additional Information](#additional-information)
+  - [Dataset Curators](#dataset-curators)
+  - [Licensing Information](#licensing-information)
+  - [Citation Information](#citation-information)
```
```suggestion
  - [Citation Information](#citation-information)
  - [Contributions](#contributions)
```
we have a new section now ;)

lhoestq commented on datasets/ccaligned_multilingual/README.md:
```diff
@@ -0,0 +1,169 @@
+
+### Citation Information
+
+@inproceedings{elkishky_ccaligned_2020,
+ author = {El-Kishky, Ahmed and Chaudhary, Vishrav and Guzm{\'a}n, Francisco and Koehn, Philipp},
+ booktitle = {Proceedings of the 2020 Conference on Empirical Methods in Natural Language Processing (EMNLP 2020)},
+ month = {November},
+ title = {{CCAligned}: A Massive Collection of Cross-lingual Web-Document Pairs},
+ year = {2020}
+ address = "Online",
+ publisher = "Association for Computational Linguistics",
+ url = "https://www.aclweb.org/anthology/2020.emnlp-main.480",
+ doi = "10.18653/v1/2020.emnlp-main.480",
+ pages = "5960--5969"
+}
```
````suggestion
```
@inproceedings{elkishky_ccaligned_2020,
 author = {El-Kishky, Ahmed and Chaudhary, Vishrav and Guzm{\'a}n, Francisco and Koehn, Philipp},
 booktitle = {Proceedings of the 2020 Conference on Empirical Methods in Natural Language Processing (EMNLP 2020)},
 month = {November},
 title = {{CCAligned}: A Massive Collection of Cross-lingual Web-Document Pairs},
 year = {2020}
 address = "Online",
 publisher = "Association for Computational Linguistics",
 url = "https://www.aclweb.org/anthology/2020.emnlp-main.480",
 doi = "10.18653/v1/2020.emnlp-main.480",
 pages = "5960--5969"
}
```

### Contributions

Thanks to [@gchhablani](https://github.com/gchhablani) for adding this dataset.
````
use code block for citation
+ add new section :D

lhoestq reviewed (commented):
Hi, Sorry for the delay ^^'

That's awesome thanks ! Good job with the  dataset config
I left a few comments.

Also could you try to reduce the size of the dummy data for `documents-ak_GH` please ? It's currently 3.8MB and it would be awesome to have something less than 20KB

gchhablani commented:
Hi @lhoestq,

Sorry for the delay, I have added the changes from the review. For the ISO format language codes, I just selected the first two characters from the names, hoping those are correct. Let me know if you want me to verify :P

Thanks for taking the time to add such a detailed review. I'll keep all these changes in mind the next time I'm adding a dataset.

Thanks,
Gunjan

lhoestq commented on datasets/ccaligned_multilingual/README.md:
````diff
@@ -0,0 +1,321 @@
+An instance of `documents` type:
+
+```
+{'Domain': 'cjtaisangha.com', 'Source_Content': 'Activities|Search|Search|Home|News|News for education|Our Club|B.A. Students|PGD Students|M.A. Students|M.Phil Studensts|Ph.D. Students|List of CJ Students|Dhamma|MP3|MP-4|Activities|CJTS.Members|E-Books|Ceylon Journey News|SGS Libaray|book list|English|Thai|Shan|Myanmar|Download|Gallery|Contact us|Donate|Links|Skip to content|Home|News|News for ed ...
````
Can you update this example now that we're using the `Translation` feature type please ?

gchhablani replied:
Yes. I missed the README. Sorry.

lhoestq commented on datasets/ccaligned_multilingual/README.md:
````diff
@@ -0,0 +1,321 @@
+An instance of `sentences` type:
+
+```
+{'LASER_similarity': 1.2734256982803345, 'Source_Sentence': '>>> PhD Students in 2018', 'Target_Sentence': '>>> လုၵ်ႈႁဵၼ်းၸၼ်ႉ Ph.D ၼႂ်းပီ 2018', 'from_english': True}
````
same here

lhoestq commented on datasets/ccaligned_multilingual/README.md:
```diff
@@ -0,0 +1,321 @@
+- `Source_URL`: a `string` feature containing the source URL.
+- `Source_Content`: a `string` feature containing the content on Source_URL.
+- `Target_URL`: a `string` feature containing the target URL.
+- `Target_Content`: a `string` feature containing the content on Target_URL.
```
The `Source_Content` and `Target_Content ` fields are now replaced by a field `translation` with two subfields.
One subfield is `en_XX` and the other one is the other language code.

lhoestq commented on datasets/ccaligned_multilingual/README.md:
```diff
@@ -0,0 +1,321 @@
+For `sentences` type:
+
+- `Source_Sentence`: a `string` feature containig the source sentence.
+- `Target_Sentence`: a `string` feature containing the target sentence.
```
same here

lhoestq reviewed (commented):
Thanks for the changes :)

I added just a few comments in the dataset card to take into account the change to the Translation feature type.

Also it looks like the dummy_data.zip file for the ak_GH configuration is quite big (3MB), can you try to reduce its size ? Ideally it should be just a few KB like for the other configs.

gchhablani commented:
Hi @lhoestq,
Earlier, I had just copied the entire `tsv.xz` file for each config. I am unable to create dummy files manually. I use `unxz` command to unzip, cut the file to 5 files manually, and zip using `xz` command.

Here is the error I receive on testing :
```sh
===================== test session starts ======================
platform linux -- Python 3.7.9, pytest-6.2.2, py-1.10.0, pluggy-0.13.1
rootdir: /home/crocoder/Desktop/Side Projects/datasets
plugins: xdist-2.2.0, forked-1.3.0
collected 1 item                                               

tests/test_dataset_common.py F                           [100%]

=========================== FAILURES ===========================
_ LocalDatasetTest.test_load_dataset_all_configs_ccaligned_multilingual _

self = <tests.test_dataset_common.LocalDatasetTest testMethod=test_load_dataset_all_configs_ccaligned_multilingual>
dataset_name = 'ccaligned_multilingual'

    @slow
    def test_load_dataset_all_configs(self, dataset_name):
        configs = self.dataset_tester.load_all_configs(dataset_name, is_local=True)
>       self.dataset_tester.check_load_dataset(dataset_name, configs, is_local=True, use_local_dummy_data=True)

tests/test_dataset_common.py:237: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/test_dataset_common.py:171: in check_load_dataset
    try_from_hf_gcs=False,
src/datasets/builder.py:579: in download_and_prepare
    dl_manager=dl_manager, verify_infos=verify_infos, **download_and_prepare_kwargs
src/datasets/builder.py:656: in _download_and_prepare
    self._prepare_split(split_generator, **prepare_split_kwargs)
src/datasets/builder.py:977: in _prepare_split
    generator, unit=" examples", total=split_info.num_examples, leave=False, disable=not_verbose
../../../anaconda3/envs/datasets_env/lib/python3.7/site-packages/tqdm/std.py:1133: in __iter__
    for obj in iterable:
../../../.cache/huggingface/modules/datasets_modules/datasets/ccaligned_multilingual/ee824b1b77c892c7ce0435bcf2e3f75b3c6c86bf700254a213c24f6b236d55c0/ccaligned_multilingual.py:197: in _generate_examples
    for id_, row in enumerate(f):
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <encodings.utf_8.IncrementalDecoder object at 0x7fa85a5eb7d0>
input = b'\xfd7zXZ\x00\x00\x04\xe6\xd6\xb4F\x02\x00!\x01\x16\x00\x00\x00t/\xe5\xa3\xe1\xa2wX\x07]\x004\x9c\xc9\xbb\x7f\xa5w\xd...xb4s\xb8\xb6\x19c\xfeE\xd0\xcc\xe88P\x9a\xb5\xc8\xc2\xb3`\xe2\xa9_[\x05\xa4\xebK\x99\xcd$\xc1\xd9\xa1\xab\xea\xf5!\xc5'
final = False

    def decode(self, input, final=False):
        # decode input (taking the buffer into account)
        data = self.buffer + input
>       (result, consumed) = self._buffer_decode(data, self.errors, final)
E       UnicodeDecodeError: 'utf-8' codec can't decode byte 0xfd in position 0: invalid start byte

../../../anaconda3/envs/datasets_env/lib/python3.7/codecs.py:322: UnicodeDecodeError
--------------------- Captured stdout call ---------------------
Downloading and preparing dataset ccaligned_multilingual/documents-zz_TR (download: 1.00 MiB, generated: 1.00 MiB, post-processed: Unknown size, total: 2.00 MiB) to /tmp/tmp215iympu/ccaligned_multilingual/documents-zz_TR/1.0.0...
Dataset ccaligned_multilingual downloaded and prepared to /tmp/tmp215iympu/ccaligned_multilingual/documents-zz_TR/1.0.0. Subsequent calls will reuse this data.
Downloading and preparing dataset ccaligned_multilingual/sentences-zz_TR (download: 1.00 MiB, generated: 1.00 MiB, post-processed: Unknown size, total: 2.00 MiB) to /tmp/tmpte071yqd/ccaligned_multilingual/sentences-zz_TR/1.0.0...
Dataset ccaligned_multilingual downloaded and prepared to /tmp/tmpte071yqd/ccaligned_multilingual/sentences-zz_TR/1.0.0. Subsequent calls will reuse this data.
Downloading and preparing dataset ccaligned_multilingual/documents-tz_MA (download: 1.00 MiB, generated: 1.00 MiB, post-processed: Unknown size, total: 2.00 MiB) to /tmp/tmp7abfbl2j/ccaligned_multilingual/documents-tz_MA/1.0.0...
Dataset ccaligned_multilingual downloaded and prepared to /tmp/tmp7abfbl2j/ccaligned_multilingual/documents-tz_MA/1.0.0. Subsequent calls will reuse this data.
Downloading and preparing dataset ccaligned_multilingual/sentences-tz_MA (download: 1.00 MiB, generated: 1.00 MiB, post-processed: Unknown size, total: 2.00 MiB) to /tmp/tmpasivrxhk/ccaligned_multilingual/sentences-tz_MA/1.0.0...
Dataset ccaligned_multilingual downloaded and prepared to /tmp/tmpasivrxhk/ccaligned_multilingual/sentences-tz_MA/1.0.0. Subsequent calls will reuse this data.
Downloading and preparing dataset ccaligned_multilingual/documents-ak_GH (download: 1.00 MiB, generated: 1.00 MiB, post-processed: Unknown size, total: 2.00 MiB) to /tmp/tmp8r_mqo5v/ccaligned_multilingual/documents-ak_GH/1.0.0...
--------------------- Captured stderr call ---------------------
0 examples [00:00, ? examples/s]2021-02-26 17:40:09.950005: W tensorflow/stream_executor/platform/default/dso_loader.cc:60] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory
2021-02-26 17:40:09.950050: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
                                    
======================= warnings summary =======================
tests/test_dataset_common.py::LocalDatasetTest::test_load_dataset_all_configs_ccaligned_multilingual
  /home/crocoder/anaconda3/envs/datasets_env/lib/python3.7/site-packages/tensorflow/python/autograph/impl/api.py:22: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
    import imp

-- Docs: https://docs.pytest.org/en/stable/warnings.html
=================== short test summary info ====================
FAILED tests/test_dataset_common.py::LocalDatasetTest::test_load_dataset_all_configs_ccaligned_multilingual
================= 1 failed, 1 warning in 8.27s =================
```

gchhablani commented:
Hi @lhoestq,

I have changed the README, and added a single example per config. Even one example is long enough to make the files heavy. Hope that isn't an issue.

Thanks,
Gunjan

lhoestq commented on datasets/ccaligned_multilingual/ccaligned_multilingual.py:
```diff
@@ -0,0 +1,216 @@
+        """Returns SplitGenerators."""
+        my_urls = _URLs[self.config.name[:9]]
+        if self.config.name[:9] == "sentences" and self.config.language_code in reverse_mapped_sentences:
+            url = os.path.join(my_urls, self.config.language_code + "-en_XX.tsv.xz")
+            from_english = False
+        else:
+            url = os.path.join(my_urls, "en_XX-" + self.config.language_code + ".tsv.xz")
```
```suggestion
            url = my_urls + self.config.language_code + "-en_XX.tsv.xz"
            from_english = False
        else:
            url = my_urls + "en_XX-" + self.config.language_code + ".tsv.xz"
```
don't use os.path.join on urls since the separator is a backslash on windows, and not a slash.

lhoestq reviewed (approved):
Thanks for the changes !
LGTM :)

gchhablani commented:
Hi @lhoestq,

Thanks for approving.
Just out of curiosity, why is `os.path.join` a problem?
